packages/zoltan2/src/algorithms/partition/Zoltan2_AlgMultiJagged.hpp

```diff
@@ -6783,7 +6783,37 @@ bool Zoltan2_AlgMJ<Adapter>::mj_premigrate_to_subset( int used_num_ranks,
   {
     ArrayRCP<mj_gno_t> received_gnos(num_incoming_gnos);
 
+#if !defined(HAVE_TPETRACORE_CUDA)  // KDD UVM HACK FOR CUDA IS BELOW
+    // This is the correct implementation, using initial_mj_gnos_ as the 
+    // data to be sent.
     ArrayView<const mj_gno_t> sent_gnos(initial_mj_gnos_, num_local_coords_);
+#else
+    // Unfortunately, this HACK is needed to workaround BUGs in SpectrumMPI.
+    // Apparently, SpectrumMPI does not communicate UVM memory correctly 
+    // without special command-line flags that dramatically slow MPI 
+    // communications.  
+    // When Tpetra is used in the XpetraMultiVectorAdapter, 
+    // initial_mj_gnos_ is obtained from Tpetra::Map::getNodeElementList().
+    // For CUDA builds, Tpetra returns UVM memory in getNodeNumElements(),
```

Comment:
See my comments.  `Tpetra::Map::getNodeElementList` returns HostSpace memory, at least in Trilinos:master.